### PR TITLE
Add REST Importer for Jetbrains / VSCode

### PR DESCRIPTION
**Pull request by benfaerber**
## Description

A few months ago I filed a pull request to add this in but I got busy so it ended up getting closed. ([Addresses this](https://github.com/LucasPickering/slumber/issues/122)) Anyway, I'm back and decided to reintegrate it! Here is the previous closed issue with more details: https://github.com/LucasPickering/slumber/pull/205

I made the REST Parser a separate crate because it was quite complex and would be useful for other projects as well. You can check it out here: https://github.com/benfaerber/rest_parser

The old version had a Jetbrains environment loading feature which I did not include because it added a ton of complexity and not too much benefit.

Import using:
```
slumber import rest ./test_data/rest_http_bin.http ./test_data/rest_slumber.yml
```
```
slumber import vscode ./test_data/rest_http_bin.http ./test_data/rest_slumber.yml
```
```
slumber import jetbrains ./test_data/rest_http_bin.http ./test_data/rest_slumber.yml
```

## Known Risks
- Adding a new crate
- Touching the importer

## QA

Lots of unit tests in `rest.rs` and some test data.

## Checklist

- [ X] Have you read `CONTRIBUTING.md` already?
- [ X] Did you update `CHANGELOG.md`?
  - Only user-facing changes belong in the changelog. Internal changes such as refactors should only be included if they'll impact users, e.g. via performance improvement.
- [ X] Did you remove all TODOs?
  - If there are unresolved issues, please open a follow-on issue and link to it in a comment so future work can be tracked

### CODE REVIEW

**Review by LucasPickering — CHANGES_REQUESTED** [review]
Nice work! This is looking great, just some formatting/readability comments mostly. I'm going to push one commit with some small changes I made to the docs, I figured it would be easier than trying to explain.

There are some formatting and lint issues as well, they'll show up in the CI once I kick it off. 

Thanks for your work on this!

File: crates/import/src/rest.rs
[the remark this answers is not in the document]
Comment (benfaerber) [reply crates/import/src/rest.rs:1]: All converted to `try_build` or `build`

**Comment by benfaerber** [comment]
Cool, I think these changes address everything mentioned. Now I just have to figure out the linters.

File: crates/import/src/rest.rs [hunk 1]
```diff
@@ -0,0 +1,664 @@
...
+        Authentication, Chain, ChainId, ChainOutputTrim, ChainSource,
+        Collection, HasId, Method, Profile, ProfileId, Recipe, RecipeBody,
+        RecipeId, RecipeNode, RecipeTree, SelectorMode,
+    },
+    http::content_type::ContentType,
+    template::{Identifier, Template},
+};
+
+use mime::APPLICATION_JSON;
+use reqwest::header::CONTENT_TYPE;
```
Comment (LucasPickering) [inline crates/import/src/rest.rs:20]: I don't like fully importing constants or functions because I think it makes them hard to follow. Could you import to the module level and use the qualified identifiers instead? (`mime::APPLICATION_JSON` and `header::CONTENT_TYPE`)

File: crates/import/src/rest.rs [hunk 2]
```diff
@@ -0,0 +1,664 @@
...
+fn guess_is_json(
+    r_headers: &IndexMap<String, RestTemplate>,
+    variables: &RestVariables,
+) -> bool {
+    for (name, value) in r_headers {
+        if name.to_lowercase() == CONTENT_TYPE.as_str() {
+            return value.render(variables) == APPLICATION_JSON.to_string();
+        }
+    }
+    false
```
Comment (LucasPickering) [inline crates/import/src/rest.rs:149]: I think the functional form is more readable (once it's been formatted at least)
```suggestion
    r_headers.iter().any(|(name, value)| name.to_lowercase() == header::CONTENT_TYPE.as_str() && value.render(variables).as_str() == mime::APPLICATION_JSON)
```

File: crates/import/src/rest.rs [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+            Ok(t) => Some((k, t)),
+            Err(err) => {
+                error!("{err}");
+                None
+            }
+        })
+        .collect()
+}
+
+fn try_build_receipe(
```
Comment (LucasPickering) [inline crates/import/src/rest.rs:210]:
```suggestion
fn try_build_recipe(
```

File: crates/import/src/rest.rs [hunk 4]
```diff
@@ -0,0 +1,664 @@
...
+) -> IndexMap<String, Template> {
+    template_map
+        .into_iter()
+        .filter_map(|(k, v)| match try_build_slumber_template(v) {
+            Ok(t) => Some((k, t)),
+            Err(err) => {
+                error!("{err}");
+                None
+            }
+        })
```
Comment (LucasPickering) [inline crates/import/src/rest.rs:91]: There's a `ResultTraced` extension in `slumber_core` that makes this simpler:
```suggestion
        .filter_map(|(k, v)| try_build_slumber_template(v).map(|v| (k,v)).traced().ok())
```
Same goes for the other 2 spots with this pattern

**Review by LucasPickering — CHANGES_REQUESTED** [review]
Few more suggestions but it's looking good! For the formatting you can run `cargo fmt` locally, for linting `cargo clippy` will show you the errors.

**Comment by benfaerber** [comment]
There we go! That should be everything. I got the linter and formatter happy too.

**Review by LucasPickering — APPROVED** [review]
Looks great! Thank you! This will go out in 2.5.0, which will _probably_ be next week some time.
